### ISSUE TITLE
Infinite loop if class is extended (specific circumstances)

### ISSUE THREAD

**Issue by xZero707** [zendframework/zend-config#55]
- [x] I was not able to find an [open](https://github.com/zendframework/zend-config/issues?q=is%3Aopen) or [closed](https://github.com/zendframework/zend-config/issues?q=is%3Aclosed) issue matching what I'm seeing.
 - [x] This is not a question.

I'm trying to extend Zend\Config\Config and supply default config. I also add some extra methods. 

### Code to reproduce the issue

```php

<?php

namespace Zend\Config\Issue;

use Zend\Config\Config;

/**
 * Class Simple
 * @package Zend\Config\Issue
 */
class Simple extends Config
{
    /** @var int */
    private static $callsCount = 0;

    /**
     * Simple constructor.
     */
    public function __construct()
    {
        self::$callsCount++;
        echo __METHOD__ . " - Number of calls: " . self::$callsCount . PHP_EOL;

        $default = [

            'something' => [
                0 => 'I am an array',
                1 => '\Zend\Config\Config::__construct() will try to create new instance of itself',
                2 => 'And that\'s fine',
                3 => 'As long as it does "new self" instead of "new static"',
                4 => 'because static means this extending class'
            ],
            'something_else' => 'this is fine'
        ];

        // Call parent constructor \Zend\Config\Config::__construct()
        parent::__construct($default);
    }


}

```

### Expected results

Definitely not infinite loop. 

### Actual results

Infinite loop - class Simple is calling Zend\Config\Config::__construct() and Zend\Config\Config will try to create new instance of itself using "new static()", which will in turn call extending class Simple .....
If you run code I provided, you will get following output:
```
Zend\Config\Issue\Simple::__construct - Number of calls: 1
Zend\Config\Issue\Simple::__construct - Number of calls: 2
Zend\Config\Issue\Simple::__construct - Number of calls: 3
Zend\Config\Issue\Simple::__construct - Number of calls: 4
Zend\Config\Issue\Simple::__construct - Number of calls: 5
Zend\Config\Issue\Simple::__construct - Number of calls: 6
Zend\Config\Issue\Simple::__construct - Number of calls: 7
Zend\Config\Issue\Simple::__construct - Number of calls: 8
(...)
```

**Comment by webimpress** [comment]
@xZero707 I wouldn't consider it as a bug but invalid use case. Because you are defining the array n the constructor and passing it to the parent and - as you noted we are using `new static` you get the infinite loop.

I think we want to have instances of the "extended" class for all "subnodes" and that's why we use `new static`. Please see #57

**Comment by xZero707** [comment]
Oh, that explains a lot. I understand your intention now.

**Comment by weierophinney** [comment]
Closed with #57.
